### DISCUSSION TITLE
Get dictionary from python

### DISCUSSION THREAD

**Discussion by robertosante** [PyO3/pyo3#2735] [Questions] [answered]
Hey how do you actually get a dictionary from a python file I'm using from_code to run a py file but when I call the file I haven't figured out how to get the dictionary from python to rust

**Comment by robertosante** [comment]
I have the following code:

`
let from_python = Python::with_gil(|py| -> PyResult<()> {
        let app: Py<PyAny> = PyModule::from_code(py, py_app, "", "")?
        .getattr("run")?
        .into();
        let kwargs = vec![(key1, val1), (key2, val2), (key3, val3), (key4, val4), (key5, val5)];
     
        let result = app.call(py, (), Some(kwargs.into_py_dict(py)));
        println!("py: {:#?}", result?);
        Ok(())
    });`

my file py_app returns a dictionary like

` d = dict();
  d['name'] = tr.beneficiario.nombre
  d['rfc'] = tr.beneficiario.rfc
  return d`

on the print result? i get: 
`py: Py(
    0x000000010d76e400,
)`

my question is how can i extract that result to something that i can use in rust

**Comment by birkenfeld** [comment, marked as answer]
It depends. If you want to keep working with it as a Python object, you can already use its methods from PyAny, and you can get to PyDict using `result?.downcast::<PyDict>()?`.

If you want to convert everything to Rust, i.e. a HashMap, use extract ().

**Reply by robertosante** [reply]
@birkenfeld  thanks for your time!  if i would like to keep it live how can i map it to a hashmap? just use extract and then assign to a hash literal?

**Reply by birkenfeld** [reply]
Yes, you assign to a variable where you specify the desired type:
```
let rust_map: HashMap<String, String> = result?.extract()?;
```

Of course, Rust maps being strong-typed, this only works if all key and value types are the same. Otherwise you can still keep some types as Python objects, e.g. `HashMap<String, Py<PyAny>>`.

**Comment by robertosante** [comment]
@birkenfeld got it thanks
